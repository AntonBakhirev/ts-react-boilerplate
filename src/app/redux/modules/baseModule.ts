import {Action} from "redux";

export interface IAction<P, T = string, M = string> extends Action<T> {
  payload?: P;
  message?: M;
}

export interface IBaseState {
  error: string;
  loaded: boolean;
  pending: boolean;
}

// trick compiler that they are literal types
export type INVOKED = "INVOKED";
export type PENDING = "PENDING";
export type FULFILLED = "FULFILLED";
export type REJECTED = "REJECTED";

export interface IPromiseActions {
  INVOKED: INVOKED;
  PENDING: PENDING;
  FULFILLED: FULFILLED;
  REJECTED: REJECTED;
}

export function getPromiseActionType(baseAction: string): IPromiseActions {
  return {
    FULFILLED: baseAction + "_FULFILLED" as FULFILLED,
    INVOKED: baseAction as INVOKED,
    PENDING: baseAction + "_PENDING" as PENDING,
    REJECTED: baseAction + "_REJECTED" as REJECTED
  };
}

export interface IActionCreators<I, P, F, R> {
  actionTypes: IPromiseActions;
  invoke: (payload: I) => IAction<I, INVOKED>;
  setPending: (payload: P) => IAction<P, PENDING>;
  setFulfilled: (payload: F) => IAction<F, FULFILLED>;
  setRejected: (payload: R) => IAction<null, REJECTED, R>;
}

export function getPromiseAction<I, P, F, R>(type: string): IActionCreators<I, P, F, R> {
  return {
    actionTypes: getPromiseActionType(type),
    invoke: (payload: I): IAction<I, INVOKED> => {
      return {type: type as INVOKED, payload};
    },
    setFulfilled: (payload: F): IAction<F, FULFILLED> => {
      return {type: getPromiseActionType(type).FULFILLED, payload};
    },
    setPending: (payload: P): IAction<P, PENDING> => {
      return {type: getPromiseActionType(type).PENDING, payload};
    },
    setRejected: (message: R): IAction<null, REJECTED, R> => {
      return {type: getPromiseActionType(type).REJECTED, message};
    }
  };
}
