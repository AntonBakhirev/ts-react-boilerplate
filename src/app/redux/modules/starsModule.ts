import {IStars, IStarsAction} from "models/starsModel";
import promiseReducer from "../../helpers/promiseReducer";

/** Action Types */
export const STARS_REQUEST: string = "stars/STARS_REQUEST";
export const STARS_SUCCESS: string = "stars/STARS_SUCCESS";
export const STARS_FAILURE: string = "stars/STARS_FAILURE";

/** Initial State */
const initialState: IStars = {
  isFetching: false
};

/** Reducer */
export function starsReducer(state: IStars = initialState, action: IStarsAction): IStars {
  return promiseReducer<IStars, IStarsAction>("stars/STARS", state, action);
}
