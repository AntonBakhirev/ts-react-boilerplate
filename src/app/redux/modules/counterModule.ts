import {IAction, IState} from "./baseModule";

export const INCREMENT: string = "counter/INCREMENT";
export const DECREMENT: string = "counter/DECREMENT";

export interface ICounter {
  count: number;
}

/** Counter: Initial State */
const initialState: IState<ICounter> = {
  isFetching: false,
  payload: {
    count: 0
  }
};

/** Reducer: CounterReducer */
export function counterReducer(state: IState<ICounter> = initialState, action?: IAction<ICounter>): IState<ICounter> {
  switch (action.type) {
    case INCREMENT:
      return {
        ...state,
        payload: {
          count: state.payload.count + 1
        }
      };

    case DECREMENT:
      return {
        ...state,
        payload: {
          count: ((state.payload.count - 1 > 0) ? state.payload.count - 1 : 0)
        }
      };

    default:
      return state;
  }
}
