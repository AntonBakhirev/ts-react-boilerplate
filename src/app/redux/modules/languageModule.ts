import promiseReducer, {IRequestType} from "../../helpers/promiseReducer";
import {IAction, IState} from "./baseModule";

export const SET_LANGUAGE = "language/SET_LANGUAGE";
export const SWITCH_LANGUAGE = "language/SWITCH_LANGUAGE";
export const requestType: IRequestType = {
  FAILURE: "language/FAILURE",
  PENDING: "language/PENDING",
  SUCCESS: SET_LANGUAGE
};

export interface ILanguage {
  languageData: any;
  locale: string;
}

const initialState: IState<ILanguage> = {
  isFetching: false,
  payload: {
    languageData: {},
    locale: "en-GB"
  }
};

export function languageReducer(state: IState<ILanguage> = initialState, action: IAction<ILanguage>): IState<ILanguage> {
  return promiseReducer<ILanguage>(requestType, state, action);
}
