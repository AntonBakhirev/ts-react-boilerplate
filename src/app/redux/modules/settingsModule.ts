import {FULFILLED, getPromiseAction, IAction, IBaseState, INVOKED, PENDING, REJECTED} from "./baseModule";

export const changeLanguage = getPromiseAction<string, null, any, string>("settings/CHANGE_LANGUAGE");

export interface ITranslations {
  [key: string]: string;
}

export interface ISettingsState extends IBaseState {
  language: string;
  translations: ITranslations;
}

const initialState: ISettingsState = {
  error: "",
  language: "en-US",
  pending: false,
  translations: {}
};

type CHANGE_LANGUAGE_PENDING = PENDING;
type CHANGE_LANGUAGE_FULFILLED = FULFILLED;
type CHANGE_LANGUAGE_REJECTED = REJECTED;

export function settingsReducer(
  state: ISettingsState = initialState,
  action: IAction<null, CHANGE_LANGUAGE_PENDING> |
    IAction<ITranslations, CHANGE_LANGUAGE_FULFILLED> |
    IAction<null, CHANGE_LANGUAGE_REJECTED, string>
): ISettingsState {
  switch (action.type) {
    case changeLanguage.actionTypes.PENDING:
      return {
        ...state,
        pending: true
      };
    case changeLanguage.actionTypes.FULFILLED:
      return {
        ...state,
        translations: action.payload
      };
    case changeLanguage.actionTypes.REJECTED:
      return {
        ...state,
        error: action.message,
        pending: false
      };
    default:
      return state;
  }
}

export function invokeChangeLanguage(language: string): IAction<string, INVOKED> {
  return {
    payload: language,
    type: changeLanguage.actionTypes.INVOKED
  };
}
