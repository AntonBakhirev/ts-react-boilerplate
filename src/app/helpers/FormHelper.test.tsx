import {shallow} from "enzyme";
import * as React from "react";
import {IState} from "../redux/modules/baseModule";
import {ILanguage} from "../redux/modules/languageModule";
import {
aol, email, mapStateToProps, matchedPwd, maxLength, minLength,
minValue, numberType, required, tooOld, UnconnectedCustomField
} from "./FormHelper";

describe("FormHelper", () => {

  describe("<CustomField />", () => {
    it("matches snapshot without error and warning", () => {
      const props = {
        defaultMessage: "Username",
        languageData: {
          username: "Username"
        },
        languageId: "username",
        meta: {
          active: false,
          error: false,
          touched: false,
          warning: false
        },
        type: "text"
      };
      const component = shallow(<UnconnectedCustomField {...props} />);
      expect(component).toMatchSnapshot();
    });

    it("matches snapshot with error when active or touched", () => {
      const props = {
        defaultMessage: "Username",
        languageData: {
          username: "Username"
        },
        languageId: "username",
        meta: {
          active: true,
          error: {id: "error", defaultMessage: "some error occurs"},
          touched: true,
          warning: false
        },
        type: "text"
      };
      const component = shallow(<UnconnectedCustomField {...props} />);
      expect(component).toMatchSnapshot();
    });

    it("matches snapshot with warning when active or touched", () => {
      const props = {
        defaultMessage: "Username",
        languageData: {
          username: "Username"
        },
        languageId: "username",
        meta: {
          active: true,
          error: false,
          touched: true,
          warning: {id: "warning", defaultMessage: "something needs to be warned"}
        },
        type: "text"
      };
      const component = shallow(<UnconnectedCustomField {...props} />);
      expect(component).toMatchSnapshot();
    });

    it("matches snapshot when placeholder falls back to defaultMessage when languageData key does not exist", () => {
      const props = {
        defaultMessage: "Username",
        languageData: {},
        languageId: "username",
        meta: {
          active: false,
          error: false,
          touched: false,
          warning: false
        },
        type: "text"
      };
      const component = shallow(<UnconnectedCustomField {...props} />);
      expect(component).toMatchSnapshot();
    });

    it("maps state to props correctly", () => {
      const language: IState<ILanguage> = {
        payload: {
          languageData: {greeting: "Hello!"},
          locale: "en-GB"
        }
      };
      const props = mapStateToProps({language});
      expect(props.languageData).toEqual(language.payload.languageData);
    });
  });

  describe("aol()", () => {
    it("invalidates aol email", () => {
      expect(aol("aol", "Really? You still use AOL for your email?")("pai@aol.com")).toEqual({
        defaultMessage: "Really? You still use AOL for your email?",
        id: "aol"
      });
    });

    it("validates email from other domains", () => {
      expect(aol("aol" , "Really? You still use AOL for your email?")("pai@pai.com")).toBeUndefined();
    });
  });

  describe("email()", () => {
    it("invalidates non email", () => {
      expect(email("invalidemail", "Invalid email format")("abcd")).toEqual({
        defaultMessage: "Invalid email format",
        id: "invalidemail"
      });
    });
    it("validates email correctly", () => {
      expect(email("invalidemail", "Invalid email format")("pai@pai.com")).toBeUndefined();
    });
  });

  describe("matchedPwd()", () => {
    it("invalidates unmatched passwords", () => {
      expect(matchedPwd("pwdunmatched", "Passwords not matched")("pass1", {password: "pass2"})).toEqual({
        defaultMessage: "Passwords not matched",
        id: "pwdunmatched"
      });
    });
    it("validates matched passwords", () => {
      expect(matchedPwd("pwdunmatched", "Passwords not matched")("pass", {password: "pass"})).toBeUndefined();
    });
  });

  describe("maxLength()", () => {
    it("invalidates too long string", () => {
      expect((maxLength("maxchar", "Must be {max} characters or less")(8)("123456789"))).toEqual({
        defaultMessage: "Must be {max} characters or less",
        id: "maxchar",
        values: {
          max: 8
        }
      });
    });
    it("validates ok string", () => {
      expect((maxLength("maxchar", "Must be {max} characters or less")(8)("12345678"))).toBeUndefined();
    });
  });

  describe("minLength()", () => {
    it("invalidates too short string", () => {
      expect((minLength("minchar", "Must be {min} characters or more")(8)("1234567"))).toEqual({
        defaultMessage: "Must be {min} characters or more",
        id: "minchar",
        values: {
          min: 8
        }
      });
    });
    it("validates ok string", () => {
      expect((minLength("minchar", "Must be {min} characters or more")(8)("12345678"))).toBeUndefined();
    });
  });

  describe("minValue()", () => {
    it("invalidates too little value", () => {
      expect(minValue("minvalue", "Must be at least {min}")(5)(0)).toEqual({
        defaultMessage: "Must be at least {min}",
        id: "minvalue",
        values: {
          min: 5
        }
      });
    });

    it("validates ok value", () => {
      expect(minValue("minvalue", "Must be at least {min}")(5)(10)).toBeUndefined();
    });
  });

  describe("numberType()", () => {
    it("invalidates non number", () => {
      expect(numberType("number", "Must be a number")("abcd")).toEqual({
        defaultMessage: "Must be a number",
        id: "number"
      });
    });

    it("validates number", () => {
      expect(numberType("number", "Must be a number")(5)).toBeUndefined();
    });
  });

  describe("required()", () => {
    it("invalidates falsy value", () => {
      expect(required("required", "This is required")(0)).toEqual({
        defaultMessage: "This is required",
        id: "required"
      });
    });

    it("validates truthy value", () => {
      expect(required("required", "This is required")("abcd")).toBeUndefined();
    });
  });

  describe("tooOld()", () => {
    it("invalidates number more than 65", () => {
      expect(tooOld("tooold", "You are too old for this")(70)).toEqual({
        defaultMessage: "You are too old for this",
        id: "tooold"
      });
    });

    it("validates number less than 65", () => {
      expect(tooOld("tooold", "You are too old for this")(65)).toBeUndefined();
    });
  });

});
