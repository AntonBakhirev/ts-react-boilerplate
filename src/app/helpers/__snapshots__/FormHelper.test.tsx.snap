// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormHelper <CustomField /> placeholder falls back to defaultMessage when languageData key does not exist 1`] = `
<div
  className="fyjcn3x"
>
  <FormattedMessage
    defaultMessage="Username"
    id="username"
    values={Object {}}
  />
  <div>
    <input
      className="f17g2th6"
      placeholder="Username"
      type="text"
    />
  </div>
</div>
`;

exports[`FormHelper <CustomField /> with error when active or touched 1`] = `
<div
  className="fyjcn3x"
>
  <FormattedMessage
    defaultMessage="Username"
    id="username"
    values={Object {}}
  />
  <div>
    <input
      className="f17g2th6"
      placeholder="Username"
      type="text"
    />
    <div
      className="f1w4o4fu"
    >
      <span
        className="f1jvcvsh"
      >
        <FormattedMessage
          defaultMessage="some error occurs"
          id="error"
          values={Object {}}
        />
      </span>
    </div>
  </div>
</div>
`;

exports[`FormHelper <CustomField /> with warning when active or touched 1`] = `
<div
  className="fyjcn3x"
>
  <FormattedMessage
    defaultMessage="Username"
    id="username"
    values={Object {}}
  />
  <div>
    <input
      className="f17g2th6"
      placeholder="Username"
      type="text"
    />
    <div
      className="f1w4o4fu"
    >
      <span
        className="fr6gy5q"
      >
        <FormattedMessage
          defaultMessage="something needs to be warned"
          id="warning"
          values={Object {}}
        />
      </span>
    </div>
  </div>
</div>
`;

exports[`FormHelper <CustomField /> without error and warning 1`] = `
<div
  className="fyjcn3x"
>
  <FormattedMessage
    defaultMessage="Username"
    id="username"
    values={Object {}}
  />
  <div>
    <input
      className="f17g2th6"
      placeholder="Username"
      type="text"
    />
  </div>
</div>
`;
